File: cppyy/calib.py
#!/usr/bin/env python3.10

### CPPYY ###

from refspec import SpecConfig, PowerSpecSource, CombSource, SignalCombiner, SpecOutput, RefSpectrometer

from cppyy.gbl.std import vector

import argparse

#######################################
parser = argparse.ArgumentParser()
parser.add_argument("-v", "--verbose",      help="Verbose mode",                    action='store_true')

parser.add_argument("-c", "--cal_signal",   help="Cal signal",                      action='store_true')
parser.add_argument("-p", "--pf_signal",    help="PF signal",                       action='store_true')
parser.add_argument("-s", "--sky_signal",   help="Sky signal",                      action='store_true')
parser.add_argument("-n", "--notch",        help="Spectral notch",                  action='store_true')

parser.add_argument("-R", "--ampl_resc",    help="Calibration Amplitude Rescaling", type=float, default=1.0)
parser.add_argument("-A", "--ADC_range",    help="ADC range peak to peak",          type=float, default=2.0)

parser.add_argument("-N", "--Ngo",          help="Number of Big Samples",           type=int,   default=120)
parser.add_argument("-a", "--a2",           help="Average 2",                       type=int,   default=20)

parser.add_argument("-d", "--drift",        help="Clock drift in ppm",              type=float, default=0.0)

parser.add_argument("-P", "--pk_fname",     help="Filename for power spectrum",     type=str,   default='data/Pk/Pk_wnoise.txt')
parser.add_argument("-C", "--cal_fname",    help="Filename for calibrator",         type=str,   default='data/samples/calib_filt.txt')
parser.add_argument("-F", "--picket_fname", help="Filename for picket fence",       type=str,   default='data/samples/picket_fence_1024.txt')



parser.add_argument("-r", "--shift_rec",    help="No.records to shift calibrator",  type=int,   default=0)

parser.add_argument("-o", "--out_root",     help="Output root",                     type=str,   default='')
parser.add_argument("-W", "--write_to",     help="Do not run spectrometer, instead write to file int16 samples", type=str,   default='')
parser.add_argument("-S", "--seed",         help="Random seed",                     type=int,   default=100)

#######################################


args        = parser.parse_args()

verbose     = args.verbose

pf_signal   = args.pf_signal
sky_signal  = args.sky_signal
cal_signal  = args.cal_signal

notch       = args.notch
cal_A       = args.ampl_resc
ADC_range   = args.ADC_range

Ngo         = args.Ngo
a2          = args.a2
drift       = args.drift

pk_fname    = args.pk_fname
cal_fname   = args.cal_fname
picket_fname= args.picket_fname

cal_shift   = args.shift_rec
out_root    = args.out_root
write_to    = args.write_to
seed        = args.seed

# -----------------------------------------------------------------------------------------
if verbose: print("*** Verbose mode ***")

if (not (pf_signal or sky_signal or cal_signal)): # if nothing specified, assume all signals are set
    pf_signal=sky_signal=cal_signal=True
    if verbose: print('*** No signals specified, assuming all ***')

# Parameter printout/check:
if verbose:
    print(f'''*** PF Signal: {pf_signal}, Sky Signal: {sky_signal}, Cal Signal: {cal_signal}, Spectral Notch: {notch}, Amplitude Rescaling: {cal_A}, ADC range: {ADC_range} ***''')
    print(f'''*** Number of big samples: {Ngo}, Average 2: {a2}, Drift (ppm): {drift} ***''')
    print(f'''*** Power spectrum filename: {pk_fname}, Calibrator filename: {cal_fname}, Output root: {out_root}, Write to: {write_to} ***''')

# -----------------------------------------------------------------------------------------
cfg = SpecConfig() # the configuration object

cfg.Nfft            = 4096
cfg.Ntaps           = 6
cfg.Nchannels       = 1
cfg.Average1Size    = 50
cfg.Average2Size    = a2
cfg.notch           = notch
cfg.Ncalib          = 1024
cfg.calib_odd       = True

cfg.sanity_check()

fundamental         = cfg.fundamental_frequency()
block_size          = cfg.Nfft
Nbins               = cfg.Nbins()

if verbose: print("*** Nbins:", Nbins, "   Ncalib:", cfg.Ncalib, "***")

# Mimic the file names in the C++ source

of      = open(out_root+"powspect.txt", 'w')
ofc     = open(out_root+"calib.txt", 'w')
ofc2    = open(out_root+"calib_meta.txt", 'w')

slist = [] # array of sources

if sky_signal:
    SigNoise = PowerSpecSource(pk_fname, cfg.sampling_rate, block_size, cfg.Nchannels, Ngo*cfg.AverageSize()+cfg.Ntaps, False, False, seed)
    SigNoise.set_verbose(verbose) #   print(SigNoise.get_verbose())
    slist.append(SigNoise)
    if verbose: print ("*** Added Sky Signal ***")

if pf_signal:
    PF = CombSource(block_size, cfg.Nchannels, 1024, picket_fname, 1, 0.008)
    slist.append(PF)
    if verbose: print ("*** Added PF Signal ***")

if cal_signal:
    CalSig = CombSource(block_size, cfg.Nchannels, 2048, cal_fname, 16, cal_A, 0.0, drift*1e-6,0.0, cal_shift)
    slist.append(CalSig)
    if verbose: print ("*** Added Cal Signal ***")

# -- Combine all of the optional sources created above:
source =  SignalCombiner(slist, True)

# ---
# Check the "write only" option:
if write_to!='':
    Nwrite = int(cfg.Average1Size*cfg.Average2Size*Ngo*cfg.Nfft/block_size)
    if verbose: print(f'''*** Dumping {Nwrite} blocks of {block_size} int16 samples to {write_to} ***''')
    of = open(write_to, 'wb')
    
    minval = +32767
    maxval = -32767
    
    minval_f = +1e30
    maxval_f = -1e30

    source.init_internal(cfg.Nchannels)
    if verbose: print(f'''*** Initialized the internal buffer of {cfg.Nchannels} elements ***''')

    for i in range(0, Nwrite):
        source.next_block()
        for j in range(0, block_size):
            val_f = source.get_internal(0, j)
            minval_f = min(minval_f, val_f)
            maxval_f = max(maxval_f, val_f)
            val = int(val_f*65536/ADC_range)
            of.write((int(val)).to_bytes(2, byteorder='big', signed=True))

    of.close()
    if verbose:
        print(f'''*** Min val float = {minval_f}   Max val float = {maxval_f} ***''')
        print(f'''*** Min val = {minval}   Max val = {maxval} ***''')
        print("*** Done. Exiting. ***")

    exit(0)

if verbose: print("*** Signal Combiner Source created ***")

output          =   SpecOutput(cfg)
if verbose: print("*** Output Object created ***")

spectrometer    =   RefSpectrometer(source, cfg)
if verbose: print("*** Spectrometer created ***")


for iNgo in range(Ngo):
    spectrometer.run(output)
    if(verbose): print("*** Source rms: ", source.rms())

    # newlist = [str(x) for x in output.avg_pspec[0]]    
    #for i in range(1, Nbins):
        
        # newlist = [str(x) for x in output.avg_pspec[0]]
        # print(newlist)
        # print(" ".join(newlist))

        # print(output.avg_pspec[0][i])
        # This following line in the C++ source but it's unclear why: k = fundamental*float(i)*(10**(-6))
       


of.close()
ofc.close()
ofc2.close()

exit(0)

File: calibrator_review22/calib.cpp
#include "RefSpectrometer.h"
#include "PowerSpecSource.h"
#include "CombSource.h"
#include "WhiteNoise.h"
#include "SignalCombiner.h"
#include "SpecConfig.h"
#include "SpecOutput.h"
#include <math.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <iomanip>

#include<fftw3.h>
#include "lyra.hpp"


double PkFunc (double k) {
  return 1e-6+1e-5*exp(-pow(k-25,2)/(2*3*3)); //Pk is in dvar / Hz
}


int main(int argc, char *argv[]) {
  //fftw shit.
  //fftwf_init_threads();
  //fftwf_plan_with_nthreads(8);
  
  SpecConfig cfg;
  bool sky_signal, pf_signal, cal_signal;
  double cal_A = 1.0;
  double drift = 0.0;
  double ADC_range = 2;
  
  size_t Ngo          = 120;
  size_t cal_shift    = 0;
  int seed            = 100;

  cfg.Nfft            = 4096;
  cfg.Ntaps           = 6;
  cfg.Nchannels       = 1;
  cfg.Average1Size    = 16; // now set by Nnotch
  cfg.Average2Size    = 1;
  cfg.notch           = true;
  cfg.Ncalib          = 0; // 1024; now we are doing it the new way
  cfg.calib_odd       = true;

  std::string pk_fname="data/Pk/Pk_wnoise.txt";
  std::string cal_fname="data/samples/calib_filt.txt";
  std::string out_root="";
  std::string write_to="";
  bool help;
  auto cli = lyra::cli()
    | lyra::opt(sky_signal)
    ["-s"]["--sky"]("sky signal")
    | lyra::opt(pf_signal)
    ["-p"]["--pf"]("pf signal")
    | lyra::opt(cal_signal)
    ["-c"]["--cal"]("cal signal")
    | lyra::opt (cfg.notch)
    ["-n"]["--notch"]("spectral Notch")
    | lyra::opt (cal_A, "calibrator rescaling")
    ["-A"]["--ampl"]("Calibration amplitude rescaling")
    | lyra::opt (Ngo, "N_go")
    ["-N"]["--Ngo"]("Number of big samples")
    | lyra::opt (cfg.Average2Size, "Average 2")
    ["--a2"]("Average 2 size")
    | lyra::opt (drift, "drift_ppm")
    ["-d"]["--drift"]("clock drift in ppm")
    | lyra::opt (pk_fname, "pk_fname")
    ["--pkfn"]("filename for power spectrum")
    | lyra::opt (cal_fname, "cal_fname")
    ["--calfn"]("filename for calibrator")
    | lyra::opt (cal_shift, "cal_shift")
    ["--calshift"]("number of records to shift calibrator by")
    | lyra::opt (out_root, "out_root")
    ["--outroot"]("output_root")
    | lyra::opt (seed, "seed")
     ["--seed"]("random seed")
    | lyra::opt (write_to, "write_to")
     ["-W"]("Do not run spectrometer, instead write to file int16 samples")
    | lyra::opt (ADC_range, "ADC_range")
     ["--ADC"]("ADC range peak to peak")
    | lyra::help(help)

    ;

  auto result = cli.parse( { argc, argv } );
  if ( !result ) {
    std::cerr << "Error in command line: " << result.errorMessage() << std::endl;
    exit(1);
  }

  if(help) {
    std::cout << cli << std::endl;
    exit(0);
  }

  if (! (sky_signal || pf_signal || cal_signal)) {
    std::cout << "No signals, assuming all." << std::endl;
    sky_signal = pf_signal = cal_signal = true;
  }

  cfg.sanity_check();

  double fundamental  = cfg.fundamental_frequency();

  size_t block_size   = cfg.Nfft;

  //WhiteNoise Noise (block_size,cfg.Nchannels,1000.0);

  std::vector<SignalSource*> slist;

  PowerSpecSource* SigNoise;
  if (sky_signal){
  SigNoise = new PowerSpecSource (pk_fname, cfg.sampling_rate,
				  block_size, cfg.Nchannels, Ngo*cfg.AverageSize()+cfg.Ntaps,
				  false, false, seed);
  slist.push_back(SigNoise);
  }

  CombSource* PF;
  if (pf_signal) {
  PF = new CombSource(block_size, cfg.Nchannels, 1024,
		"data/samples/picket_fence_1024.txt", 1, 0.008);
  slist.push_back(PF);
  }

  CombSource* CalSig;
  if (cal_signal) {
  CalSig = new CombSource(block_size, cfg.Nchannels, 2048,
			  cal_fname,16, cal_A,0.0,drift*1e-6,0.0, cal_shift);
  slist.push_back(CalSig);
  }

  
  SignalCombiner source(slist, true);

  if (write_to!="") {
    float **buffer = new float*[cfg.Nchannels];
    size_t Nwrite = cfg.Average1Size*cfg.Average2Size*Ngo*cfg.Nfft/block_size;
    float buf;
    std::cout << "Dumping "<<Nwrite<<" blocks of "<<block_size<<" int16 samples to "<<write_to <<std::endl;
    std::ofstream of(write_to, std::ios::binary);
    int16_t minval,maxval;
    float minval_f, maxval_f;
    minval = +32767;
    maxval = -32767;
    minval_f = +1e30;
    maxval_f = -1e30;

    for (size_t i=0;i<Nwrite;i++) {
      source.next_block(buffer);
      for (size_t j=0;j<block_size;j++) {
	float val_f = buffer[0][j];
	minval_f = std::min(minval_f,val_f);
	maxval_f = std::max(maxval_f,val_f);
	int16_t val = int16_t(val_f*65536/ADC_range);
	minval = std::min(minval,val);
	maxval = std::max(maxval,val);
	of.write(reinterpret_cast<char*>(&val),sizeof(int16_t));
      }
    }
    of.close();
    std::cout << "Min val float = "<<minval_f << " Max val float = " <<maxval_f << std::endl;
    std::cout << "Min val = "<<minval << " Max val = " <<maxval << std::endl;
    std::cout << "Done. Exiting." << std::endl;
    return 0;
  }
		  
  
 
  SpecOutput O(&cfg);
  RefSpectrometer S(&source,&cfg);
  S.print_info();

  std::ofstream of(out_root+"powspec.txt");
  std::ofstream ofc(out_root+"calib.txt");
  std::ofstream ofc2(out_root+"calib_meta.txt");
  std::ofstream ofn(out_root+"notch.txt");

  of<<std::setprecision(16);
  ofc<<std::setprecision(16);
  ofc2<<std::setprecision(16);
  ofn<<std::setprecision(16);


  
  for (size_t i=0;i<Ngo;i++) {
    std::cout <<"--------" <<std::endl;
    S.run(&O);

    std::cout << "RMS : ";
    for (double v : source.rms()) std::cout <<v <<" ";
    std::cout << std::endl;
  

    for (size_t i=1;i<cfg.Nbins();i++) {
      double k = fundamental*i/1e6;
      of << O.avg_pspec[0][i] << " ";
    }
    of << std::endl; 

    if (cfg.Ncalib>0) {
      for (size_t i=0;i<cfg.Ncalib;i++) {
        ofc <<  O.calib_out[0][i] << " ";
      }
      ofc << std::endl;
    
    ofc2 << O.calib_drift_count << " " <<O.calib_det <<std::endl;
    }

    if (cfg.notch) {
      for (size_t i=1;i<cfg.Nbins();i++) {
        std::complex<float> v = O.get_notch_out(0,i);
        if (i==2) std::cout << v.real() << " " << v.imag() << std::endl;
        ofn << v.real() << " " << v.imag() << " ";
      }  
      ofn << std::endl;
    }

  }
  
  of.close();
  ofc.close();
  ofc2.close();
  ofn.close();

  if (sky_signal) delete SigNoise;
  if (pf_signal) delete PF;
  if (cal_signal) delete CalSig;
  return 0;
}
